### PR TITLE
Fixed Fixed Point Point

### PR DESCRIPTION
**Pull request by mjmagee991**
Fixed the PWM equations to work with fixed point correctly

### CODE REVIEW

**Review by AarC10 — APPROVED** [review]

**Comment by AarC10** [comment]
Seems good to me. May I ask what was the issue with the autoreload?

**Comment by mjmagee991** [comment]
> Seems good to me. May I ask what was the issue with the autoreload?

Before you changed it, the equation was `autoReload = period * clockFrequency / (prescaler + 1) - 1;`
After you changed it, the equation was `autoReload = clockFrequency / (prescaler + 1) - 1 / period;`

This is still algebraically correct, but C++ applies integer division to any expression where two integers are involved, which basically means it always rounds to an integer. Because period is always > 0, 1 / period will always evaluate to either 1 or 0, which was the problem. 

It also seems that you didn't account for the fact that period is now in microseconds instead of seconds, so you have to cancel that out by dividing the first term by 1000000 (microseconds/second).

The solution I came up with was `autoReload = period * (uint64_t) clockFrequency / (prescaler + 1) / 1000000 - 1;`
This takes the value `period * clockFrequency / (prescaler + 1)` from the original equation and divides by 10^6 to counteract the  fact that it is now in microseconds. I also cast to `uint64_t` because `period * clockFrequency` could overflow 32 bits when `period` is in microseconds, which would cause a problem.

**Comment by AarC10** [comment]
Alright thanks for the explanation. Great work!

**Review by zsteig — APPROVED** [review]

File: samples/pwm/main.cpp [hunk 1]
```diff
@@ -14,11 +14,37 @@ int main() {
...
 
     IO::PWM& pwm = IO::getPWM<IO::Pin::PC_0>();
+    IO::UART& uart = IO::getUART<IO::Pin::UART_TX, IO::Pin::UART_RX>(9600);
     // 1000000 microseconds (1 second) period
-    pwm.setPeriod(1000000);
+    pwm.setPeriod(100000);
     // 50 % duty cycle
     pwm.setDutyCycle(50);
 
+    uint32_t autoReload = 0;
```
Comment (cbolles) [inline samples/pwm/main.cpp:23]: You probably want to remove this test code, or if you wanted to keep it, move it to its own test.

**Review by cbolles — CHANGES_REQUESTED** [review]

File: samples/pwm/main.cpp (hunk 1 above)
Comment (mjmagee991) [reply samples/pwm/main.cpp]: Nice catch

**Review by cbolles — APPROVED** [review]
Looks good!

**Pull request merged**